Integration tests and improved regression tests for Eclipse files

ptaule opened the pull request:
In this PR, the comparison library files EclFilesComparator.cpp and EclFilesComparator.hpp support comparison of RFT files and also supports comparison of more keyword types (bool, int and char keyword types). Also there are some minor changes to the summary compare libraries and the applications in test_util/.

joakim-hove commented on opm/test_util/EclFilesComparator.cpp:
```diff
@@ -35,105 +35,147 @@ bool ECLFilesComparator::keywordValidForComparing(const std::string& keyword) co
     if (numCells1 != numCells2) {
         std::cout << "For keyword " << keyword << " and occurence " << occurence << ":\n";
-        OPM_THROW(std::runtime_error, "Number of active cells are different.");
+        OPM_THROW(std::runtime_error, "Number of cells are different.");
```
Here copying gives you nothing compared to `ecl_kw_equal (..)`.

joakim-hove replied:
The same applies to the string and bool based comparisons.

joakim-hove commented:
A .RFT file can contain three different types of RFTs:
1. Plain RFT.
2. PLT
3. PLT with segment information.

Opm currently only supports 1).   The ert ecl_rft_file abstracts away the difference between 1 and 2 - to the enten possible. So using that it will be easier to compare Opm and Eclipse results in the (quite common) case that Eclipse has written a PLT.

ptaule commented:
Ok, thanks for the info! This is a regression test i.e. flow vs flow, so I guess it might be beneficial to be able to compare all cases if flow should support 2. and 3. in the future(?) I will, however, look into it for the integration test. 

> -- to the enten possible.

What do you mean by this?

joakim-hove commented:
To the _extent possible_ it should be.

ptaule committed:
Added integration tests in compare libraries

- The compare library summaryIntegrationTest.cpp and header was added,
 which implements a integration test for summary files.

- The summary regression test was enhanced with extended
 functionality.

- EclFilesComparator.cpp with header file now includes an integration
 test for SGAS, SWAT and PRESSURE in UNRST files, and also the
 regression test was improved with cell volume check, as well as checks of more
 keyword types.

ptaule committed:
Changed to two applications for comparison

- restartRegressionTest, initRegressionTest was squashed to one
 application, which also supports integration test as well as test
 of restart files (non-unified) and .RFT files.

- summaryRegressionTest was renamed compareSummary -- now it can
 execute an integration test as well as regression test. Also other
 improvements were made.

ptaule commented:
This is ready for review.

joakim-hove commented:
It is not clear to me who is supposed to review and merge this - but if it is me I will look into it on monday.

Right now the Travis build is red, it is not clear to me why that is the case,  but the PR will not be merged before Travis is green - or dependencies to other PR's have been clearly explained.

ptaule commented:
I think the reason why the Travis build fails is that this patch renames the applications, which makes the ctest added here OPM/opm-simulators#754 fail. It should be fixed with OPM/opm-simulators#771 and OPM/opm-common#161 though.

joakim-hove commented:
> It is not clear to me who is supposed to review and merge this

By private communication from @alfbr this will be handled by @akva2.

joakim-hove commented on opm/test_util/EclFilesComparator.cpp:
```diff
@@ -305,3 +217,402 @@ double ECLFilesComparator::average(const std::vector<double>& vec) {
+        }
+    }
+    Deviation dev = calculateDeviations(val1, val2);
+    if (dev.abs > absTolerance && dev.rel > relTolerance) {
```
I guess this should be: `if (dev.abs > absTolerance || dev.rel > relTolerance)` - i.e. we should flag an error if _either_ the absolute or the relative deviation is outside the limits?

akva2 replied:
&& was the intention. the idea is that the absolute test 'should' fail in general, so the relative governs the general behavior. the exception is around 0 where the relative test will fail in general, the absolute test then 'runs the show'.

joakim-hove commented on opm/test_util/EclFilesComparator.cpp:
```diff
@@ -305,3 +217,402 @@ double ECLFilesComparator::average(const std::vector<double>& vec) {
+
+
+
+void RegressionTest::boolComparisonForOccurrence(const std::string& keyword, int occurrence1, int occurrence2) const {
```
The methods `RegressionTest::$typeComparisonForOccurence()` probably work just fine, but for the trivial `==` based comparisons a simpler alternatively can be implemented using the `ecl_kw_equal()` function. That way some code is shaved off - and the same implementation can be shared for `bool, char` and `int`.

akva2 replied:
i don't see how i can do the equivalent using ecl_kw_equal. it won't tell me where it differs.

joakim-hove replied:
You can use `ecl_kw_first_different`, but I guess that for the not-floating point vectors it is either all  equal - or not.

But by all means, just a suggestion on my part.

akva2 replied:
a good idea and i appreciate it, but since it wouldn't be doing the same thing i'll keep the current code for now.

akva2 commented:
replaced by #90
